src/ddmd/target.d

```diff
@@ -465,75 +441,201 @@ struct Target
+                break;
+            case Tint32:
+            case Tuns32:
+                *(cast(uint*) bp) = cast(uint) value;
+                break;
+            case Tint64:
+            case Tuns64:
+                *(cast(ulong*) bp) = cast(ulong) value;
+                break;
+            default:
+                assert (0);
+            }
         }
-    case Tfloat64:
+        else if (ele.type.isfloating())
         {
-            double* p = cast(double*)buffer;
-            *p = cast(double)e.toReal();
-            break;
+            const value = ele.toReal();
+
+            switch (ele.type.ty)
+            {
+            case Tfloat32:
+                *(cast(float*) bp) = cast(float) value;
+                break;
+            case Tfloat64:
+                *(cast(double*) bp) = cast(double) value;
+                break;
+            case Tfloat80:
+                *(cast(real*) bp) = cast(real) value;
+                bp[80/8 .. eleSize] = 0; // Clear the padding area for consistency.
+                break;
+            default:
+                assert (0);
+            }
         }
-    default:
-        assert(0);
+        else
+            assert (0);
+
+        // If host != target endian-ness, swap the byte order for bp[0 .. eleSize] here.
+        static if (false)
+        {
+            const last = eleSize - 1;
+            for (size_t x = 0; x <= last; ++x)
+                bp[x] = bp[last - x];
+        }
+
+        bp += eleSize;
     }
 }
 
-// Write the bytes encoded in 'buffer' into a real_t and returns
-// the value as a new RealExp.
-extern (C++) static Expression decodeReal(Loc loc, Type type, ubyte* buffer)
+// Reinterpret `buffer` as though it was of `type` and copy the value into a new Expression.
+extern (C++) Expression paintDecode(Loc loc, Type type, ubyte* buffer)
 {
-    real_t value;
-    switch (type.ty)
+    ubyte* bp = buffer;
+    Expressions* arr;
+    int aLen;
+    int ax = 0;
+    Expression ele = null;
+    Type eleType;
+
+    if (type.ty == Tsarray)
     {
-    case Tfloat32:
+        arr = new Expressions();
+        aLen = cast(int) (cast(TypeSArray) type).dim.toInteger();
+        arr.setDim(aLen);
+        eleType = (cast(TypeSArray) type).next;
+    }
+    else
+    {
+        arr = null;
+        aLen = 1;
+        eleType = type;
+    }
+
+    int eleSize = cast(int) eleType.size();
+    bool asInt = eleType.isintegral();
+    assert (asInt || eleType.isfloating());
+
+    while (ax < aLen)
+    {
+        assert (bp + eleSize <= buffer + Target.paintMaxSize &&
+            (cast(size_t) bp) % eleSize == 0);
+
+        // If host != target endian-ness, swap the byte order for bp[0 .. eleSize] here.
+        static if (false)
         {
-            float* p = cast(float*)buffer;
-            value = real_t(*p);
-            break;
+            const last = eleSize - 1;
+            for (size_t x = 0; x <= last; ++x)
+                bp[x] = bp[last - x];
```

Comment:
Or maybe my thoughts about endianness don't apply here, as dmd doesn't internally represent data as the target sees it.